Source/ParticleInterpolators/ParticleInterpolators.impl.hpp

```diff
@@ -0,0 +1,655 @@
+
+                // reflect into valid region for seeding
+                const amrex::Real xr = reflect_particle(
+                    static_cast<amrex::Real>(x_d[ip]), prob_lo[0], prob_hi[0],
+                    lo_reflect[0], hi_reflect[0]);
+                const amrex::Real yr = reflect_particle(
+                    static_cast<amrex::Real>(y_d[ip]), prob_lo[1], prob_hi[1],
+                    lo_reflect[1], hi_reflect[1]);
+#if AMREX_SPACEDIM == 3
+                const amrex::Real zr = reflect_particle(
+                    static_cast<amrex::Real>(z_d[ip]), prob_lo[2], prob_hi[2],
+                    lo_reflect[2], hi_reflect[2]);
+#endif
+
+                // set position
+                p.pos(0) = xr;
+                p.pos(1) = yr;
+#if AMREX_SPACEDIM == 3
+                p.pos(2) = zr;
+#endif
+                p.idata(0) = ip;
+
+                for (int s = 0; s < ncomp; ++s)
+                    particle_data.rdata(s)[ip] =
+                        0.0; // for now set all to zero (this is where we will
+                             // store the interpolated values)
+            });
+
+        amrex::Gpu::streamSynchronize();
+    }
+
+    // amrex::ParallelDescriptor::Barrier(); // TODO! test if this makes a
+    // difference.
+}
+
+// interpolate variables into SOA slots
+template <int num_components>
+void ParticleInterpolators<num_components>::interpolate_to_particle()
+{
+    AMREX_ASSERT(m_initialized);
+
+    ensure_redistributed();
+
+    const int start_comp = m_start_comp;
+    const int ncomp      = num_components;
+
+    for (int lev = 0; lev <= m_gr_amr->finestLevel(); ++lev)
+    {
+        // if particles are only on the coarsest level, skip the looping over
+        // levels
+        if (this->NumberOfParticlesAtLevel(lev) == 0)
+            continue;
+
+        amrex::AmrLevel &level      = m_gr_amr->getLevel(lev);
+        const amrex::Geometry &geom = level.Geom();
+        amrex::MultiFab &state      = level.get_new_data(static_cast<int>(0));
+
+        AMREX_ASSERT(start_comp + ncomp <= state.nComp());
+
+        amrex::IntVect nghost(AMREX_D_DECL(2, 2, 2));
+        state.FillBoundary(start_comp, ncomp, nghost, geom.periodicity());
+
+        const auto plo = geom.ProbLoArray();
+        const auto dxi = geom.InvCellSizeArray();
+
+        // loop over tiles and interpolate now
+        for (ParIterType it(*this, lev); it.isValid(); ++it)
+        {
+            auto &ptile    = this->ParticlesAt(lev, it);
+            auto ptd       = ptile.getParticleTileData();
+            const int np   = it.numParticles();
+            auto fab_array = state[it].const_array();
+
+            amrex::ParallelFor(
+                np,
+                [=] AMREX_GPU_DEVICE(int ip)
+                {
+                    auto &sp = ptd[ip];
+
+                    amrex::IntVect is_nodal = amrex::IntVect::TheZeroVector();
+                    LagrangeInterpolator<5> interp; // 4th order interpolation
```

Comment:
Using my suggestion above with `constexpr interp_order`
```suggestion
                    LagrangeInterpolator<interp_order + 1> lagrange_interp;